### PR TITLE
[SYCLomatic] Refine template type arg redeclaration

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 1]
```diff
@@ -5224,9 +5224,11 @@ void DeviceFunctionInfo::mergeTextureObjectList(
 KernelCallExpr::ArgInfo::ArgInfo(const ParmVarDecl *PVD,
                                  KernelArgumentAnalysis &Analysis,
                                  const Expr *Arg, bool Used, int Index,
-                                 KernelCallExpr *BASE)
+                                 KernelCallExpr *BASE, const ParmVarDecl *TPVD)
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]:
```suggestion
                                 KernelCallExpr *BASE, bool IsDependent)
```

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 2]
```diff
@@ -5875,8 +5877,10 @@ void KernelCallExpr::buildArgsInfo(const CallExpr *CE) {
       if (auto *ArgDRE = dyn_cast<DeclRefExpr>(Arg->IgnoreImpCasts()))
         Used = isArgUsedAsLvalueUntil(ArgDRE, CE);
       const auto FD = CE->getDirectCallee();
-      ArgsInfo.emplace_back(FD ? FD->parameters()[Idx] : nullptr, Analysis, Arg,
-                            Used, Idx, this);
+      const FunctionTemplateDecl *FTD = FD ? FD->getPrimaryTemplate() : nullptr;
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: Move to 5870

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 3]
```diff
@@ -5224,9 +5224,11 @@ void DeviceFunctionInfo::mergeTextureObjectList(
 KernelCallExpr::ArgInfo::ArgInfo(const ParmVarDecl *PVD,
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: Can we check if PVD is substtemplateparametertype?

File: clang/runtime/dpct-rt/include/dpct/memory.hpp [hunk 4]
```diff
@@ -918,6 +918,8 @@ static buffer_t get_buffer(const void *ptr) {
 template <typename dataT,
           sycl::access_mode accessMode = sycl::access_mode::read_write>
 class access_wrapper {
+  static_assert(!std::is_same_v<std::remove_reference_t<dataT>, void>,
+                "dataT cannot be void");
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/memory.hpp]:
```suggestion
```

**Review by ziranzha — APPROVED** [review]

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**